File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_cpp/code_output/SQLGenerator.cpp
#include <string>
#include <vector>
#include <map>
#include <sstream>
#include <algorithm>

class SQLGenerator {
private:
    std::string table_name;

    // Helper to join strings with delimiter
    static std::string join(const std::vector<std::string>& parts, const std::string& delimiter) {
        if (parts.empty()) return "";
        std::ostringstream oss;
        for (size_t i = 0; i < parts.size(); ++i) {
            if (i > 0) oss << delimiter;
            oss << parts[i];
        }
        return oss.str();
    }

public:
    SQLGenerator(const std::string& table_name) : table_name(table_name) {}

    std::string select(const std::vector<std::string>* fields, const std::string* condition) {
        std::string fieldsStr;
        if (fields == nullptr) {
            fieldsStr = "*";
        } else {
            fieldsStr = join(*fields, ", ");
        }
        std::string sql = "SELECT " + fieldsStr + " FROM " + table_name;
        if (condition != nullptr) {
            sql += " WHERE " + *condition;
        }
        return sql + ";";
    }

    std::string insert(const std::map<std::string, std::string>& data) {
        // std::map is already sorted by key (like TreeMap)
        std::vector<std::string> fieldNames;
        std::vector<std::string> values;
        for (const auto& entry : data) {
            fieldNames.push_back(entry.first);
            values.push_back("'" + entry.second + "'");
        }
        std::string fields = join(fieldNames, ", ");
        std::string vals = join(values, ", ");
        return "INSERT INTO " + table_name + " (" + fields + ") VALUES (" + vals + ");";
    }

    std::string update(const std::map<std::string, std::string>& data, const std::string& condition) {
        // std::map is sorted by key
        std::vector<std::string> setClauses;
        for (const auto& entry : data) {
            setClauses.push_back(entry.first + " = '" + entry.second + "'");
        }
        std::string setStr = join(setClauses, ", ");
        return "UPDATE " + table_name + " SET " + setStr + " WHERE " + condition + ";";
    }

    std::string del(const std::string& condition) {
        return "DELETE FROM " + table_name + " WHERE " + condition + ";";
    }

    std::string selectFemaleUnderAge(int age) {
        return "SELECT * FROM " + table_name + " WHERE age < " + std::to_string(age) + " AND gender = 'female';";
    }

    std::string selectByAgeRange(int minAge, int maxAge) {
        return "SELECT * FROM " + table_name + " WHERE age BETWEEN " + std::to_string(minAge) + " AND " + std::to_string(maxAge) + ";";
    }
};

File: data/derived_cot/rq2_eval/translation/r1/_converted_code/cpp_to_py/SQLGenerator.py
class SQLGenerator:

    def __init__(self, table_name):
        self.table_name = table_name

    def select(self, fields=None, condition=''):
        if fields is None:
            fields = []
        if not fields:
            fields_str = '*'
        else:
            fields_str = ', '.join(fields)
        sql = 'SELECT ' + fields_str + ' FROM ' + self.table_name
        if condition:
            sql += ' WHERE ' + condition
        sql += ';'
        return sql

    def insert(self, data):
        keys = sorted(data.keys())
        fields_str = ', '.join(keys)
        values_str = ', '.join(("'" + data[key] + "'" for key in keys))
        sql = 'INSERT INTO ' + self.table_name + ' (' + fields_str + ') VALUES (' + values_str + ');'
        return sql

    def update(self, data, condition):
        keys = sorted(data.keys())
        set_items = []
        for key in keys:
            set_items.append(f"{key} = '{data[key]}'")
        set_str = ', '.join(set_items)
        sql = 'UPDATE ' + self.table_name + ' SET ' + set_str
        if condition:
            sql += ' WHERE ' + condition
        sql += ';'
        return sql

    def delete(self, condition):
        sql = 'DELETE FROM ' + self.table_name
        if condition:
            sql += ' WHERE ' + condition
        sql += ';'
        return sql

    def select_female_under_age(self, age):
        condition = f"age < {age} AND gender = 'female'"
        return self.select(condition=condition)

    def select_by_age_range(self, min_age, max_age):
        condition = f'age BETWEEN {min_age} AND {max_age}'
        return self.select(condition=condition)